test/test_lib.py

```diff
@@ -0,0 +1,442 @@
+            42 < ops.lib._Lib(None, "bar", "alice@example.com", 1, 1)
+        with self.assertRaises(TypeError):
+            ops.lib._Lib(None, "bar", "alice@example.com", 1, 1) < 42
+
+        # these two might be surprising in that they don't raise an exception,
+        # but they are correct: our __eq__ bailing means Python falls back to
+        # its default of checking object identity.
+        self.assertNotEqual(ops.lib._Lib(None, "bar", "alice@example.com", 1, 1), 42)
+        self.assertNotEqual(42, ops.lib._Lib(None, "bar", "alice@example.com", 1, 1))
+
+    def test_lib_order(self):
+        a = ops.lib._Lib(None, "bar", "alice@example.com", 1, 0)
+        b = ops.lib._Lib(None, "bar", "alice@example.com", 1, 1)
+        c = ops.lib._Lib(None, "foo", "alice@example.com", 1, 0)
+        d = ops.lib._Lib(None, "foo", "alice@example.com", 1, 1)
+        e = ops.lib._Lib(None, "foo", "bob@example.com", 1, 1)
+
+        for i in range(20):
+            with self.subTest(i):
+                libs = [a, b, c, d, e]
+                shuffle(libs)
+                self.assertEqual(sorted(libs), [a, b, c, d, e])
+
+    def test_use_bad_args_types(self):
+        with self.assertRaises(TypeError):
+            ops.lib.use(1, 2, 'bob@example.com')
+        with self.assertRaises(TypeError):
+            ops.lib.use('foo', '2', 'bob@example.com')
+        with self.assertRaises(TypeError):
+            ops.lib.use('foo', 2, ops.lib.use)
+
+    def test_use_bad_args_values(self):
+        with self.assertRaises(ValueError):
+            ops.lib.use('--help', 2, 'alice@example.com')
+        with self.assertRaises(ValueError):
+            ops.lib.use('foo', -2, 'alice@example.com')
+        with self.assertRaises(ValueError):
+            ops.lib.use('foo', 1, 'example.com')
+
+
+@patch('sys.path', new=())
+class TestLibFunctional(TestCase):
+
+    def _mkdtemp(self) -> str:
+        tmpdir = mkdtemp()
+        self.addCleanup(rmtree, tmpdir)
+        return tmpdir
+
+    def test_use_finds_subs(self):
+        """Test that ops.lib.use("baz") works when baz is inside a package in the python path."""
+        tmpdir = self._mkdtemp()
+        sys.path = [tmpdir]
+
+        _mklib(tmpdir, "foo", "bar").write_text(dedent("""
+        LIBNAME = "baz"
+        LIBAPI = 2
+        LIBPATCH = 42
+        LIBAUTHOR = "alice@example.com"
+        """))
+
+        # autoimport would be done in main
+        ops.lib.autoimport()
+
+        # ops.lib.use done by charm author
+        baz = ops.lib.use('baz', 2, 'alice@example.com')
+        self.assertEqual(baz.LIBNAME, 'baz')
+        self.assertEqual(baz.LIBAPI, 2)
+        self.assertEqual(baz.LIBPATCH, 42)
+        self.assertEqual(baz.LIBAUTHOR, 'alice@example.com')
+
+    def test_use_finds_best_same_toplevel(self):
+        """Test that ops.lib.use("baz") works when there are two baz."""
+
+        T = namedtuple("T", "desc sameTop pkgA libA patchA pkgB libB patchB")
+
+        for t in [
```

Comment:
this is a very dense test to try and pick apart.
I trust that it is doing what you want, but I'm not sure how maintainable/extendable/understandable it is.

Reply:
aww :-(
i didn't want to have a lot of test methods that all looked the same, but if it's not readable as is I don't mind.
tests shouldn't be too clever :-)

Reply:
Would it be clearer to split it into 4 loops instead of 1 loop with explicit permutations?

for same_dir in [True, False]:
  for modules in [('foo', 'foo'), ('fooA', 'fooB')]:
    etc.

Trying to figure that out from lists of (True, 'foo', 'foo') is probably easier with nested loops.